Expand developer documentation. (#42)
department-of-veterans-affairs/developer-portal: docs/testing.md

# Testing Guide for the Developer Portal

## Manually Testing Local Changes

When you make changes locally, testing those changes will often require overriding an environment variable.

> This is done by modifying your `.env` file (no envionment-specific suffix). For example, assuming you

Reviewer: ```suggestion
This is done by modifying your `.env` file (no environment-specific suffix). For example, assuming you
```

> follow-ed the setup instructions in the README.md file, the OpenAPI documentation displayed in the portal will

Reviewer: ```suggestion
followed the setup instructions in the README.md file, the OpenAPI documentation displayed in the portal will
```

be coming from our dev environment.

If you are working on OpenAPI documentation in vets-api, you will need to point the portal at your local
instance of [vets-api](https://github.com/department-of-veterans-affairs/vets-api#base-setup). You will do
this by overriding the `REACT_APP_VETSGOV_SWAGGER_API` environment variable:
```
REACT_APP_VETSGOV_SWAGGER_API=http://localhost:3000
```

## Automated Testing

> There's a variety of automated tests that will be run when you submit a pull request. In order to streamline

Reviewer: ```suggestion
There are a variety of automated tests that will be run when you submit a pull request. In order to streamline
```

the process of reviewing changes, it would be best for you to ensure each of these test suites passes locally
before submitting a pull request.

### Unit Tests

General tests specific to a component or simple interactions between related components are handled by our
unit test suite. This uses the [jest framekwork](https://jestjs.io/).
```

> npm run-script test:unit:ci

Reviewer 2: ```suggestion
npm run-script test:unit
```

CI suppresses stdout and writes to a xml file which is probably not the easiest to read. We probably even want to suggest `npm run-script test:unit -- --watch` which will run unit tests on file change.

Author: > LGTM. Maybe we should update the Node v6+ statement, to say we only support Node v10+? Just to standardize team on the version with the longest support window.

I'm OK with that.

I was curious why it said v6+. I suspected maybe that's what was used in the build process or in the CI jobs. The `Dockerfile` used for the visual regression tests is based on a `node:8` image, which I'm assuming is node version 8. I wasn't able to track down the version used in the build process. Where is that determined?

Reviewer: The build happens in the docker container, so it's node:8. I've been running on 10 so I'm sure we could just update that.

```

### End-to-End Tests

> We also use jest, combined with [Pupeteer](https://github.com/GoogleChrome/puppeteer), to run end-to-end tests

Reviewer: ```suggestion
We also use jest, combined with [Puppeteer](https://github.com/GoogleChrome/puppeteer), to run end-to-end tests
```

for accessibility issues. **NOTE: THESE ARE CURRENTLY NOT WORKING ON MASTER.**

### Visual Regression Testing

This project uses `jest-image-snapshot` for visual regression testing. Screenshots of the app are checked into git, and regenerated whenever a change is made. If they don't match, Jenkins will report a test failure. To run these tests locally, you'll first need to build the docker image:
```

> docker build -t developer_portal .

Reviewer: ```suggestion
docker build -t developer-portal .
```

(my mistake, derek caught it yesterday(
